src/core/ParticlePropertyIterator.hpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2010-2022 The ESPResSo project
+ *
+ * This file is part of ESPResSo.
+ *
+ * ESPResSo is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ESPResSo is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef ESPRESSO_SRC_CORE_PARTICLE_EXTRACT_PROPERTIES_HPP
+#define ESPRESSO_SRC_CORE_PARTICLE_EXTRACT_PROPERTIES_HPP
+
+#include "BoxGeometry.hpp"
+#include "Particle.hpp"
+#include "ParticleRange.hpp"
+#include "grid.hpp"
+#include <boost/iterator/transform_iterator.hpp>
+#include <boost/range/iterator_range.hpp>
+#include <tuple>
+#include <utils/Vector.hpp>
+
+namespace ParticlePropertyRange {
+
+namespace {
+template <class Kernel>
+auto create_transform_range(ParticleRange const &particles, Kernel kernel) {
+  auto transform_iterator_begin =
+      boost::make_transform_iterator(particles.begin(), kernel);
+  auto transform_iterator_end =
+      boost::make_transform_iterator(particles.end(), kernel);
+  return boost::make_iterator_range<decltype(transform_iterator_begin)>(
+      transform_iterator_begin, transform_iterator_end);
+}
+} // namespace
+
+auto pos_range(ParticleRange const &particles) {
+  auto return_pos = [](Particle &p) { return p.pos(); };
+  return create_transform_range(particles, return_pos);
+}
+
+auto unfolded_pos_range(ParticleRange const &particles,
+                        BoxGeometry const &box) {
+  auto return_unfolded_pos = [&box](Particle &p) {
+    return unfolded_position(p.pos(), p.image_box(), box.length());
+  };
+  return create_transform_range(particles, return_unfolded_pos);
+}
+
+auto charge_range(ParticleRange const &particles) {
+  auto return_charge = [](Particle &p) { return p.q(); };
```

Comment:
Maybe we need to yield by reference here, but I'm not sure. The ICC algorithm for example needs to be able to modify particle charges.
```suggestion
  auto return_charge = [](Particle &p) -> double & { return p.q(); };
```

Reply:
Yes, I would also assume this to be a reference (so similar to `return_force`)